File: tests/analytics/meal-regularity.test.ts
import { describe, test, expect } from 'vitest';
import { computeMealRegularity } from '$lib/analytics/meal-regularity';

describe('computeMealRegularity', () => {
	test('same meal times daily gives high score', () => {
		const entries = [
			{ date: '2024-01-01', mealType: 'Breakfast', eatenAt: '2024-01-01T08:00:00+00:00' },
			{ date: '2024-01-01', mealType: 'Lunch', eatenAt: '2024-01-01T12:00:00+00:00' },
			{ date: '2024-01-01', mealType: 'Dinner', eatenAt: '2024-01-01T19:00:00+00:00' },
			{ date: '2024-01-02', mealType: 'Breakfast', eatenAt: '2024-01-02T08:00:00+00:00' },
			{ date: '2024-01-02', mealType: 'Lunch', eatenAt: '2024-01-02T12:00:00+00:00' },
			{ date: '2024-01-02', mealType: 'Dinner', eatenAt: '2024-01-02T19:00:00+00:00' },
			{ date: '2024-01-03', mealType: 'Breakfast', eatenAt: '2024-01-03T08:00:00+00:00' },
			{ date: '2024-01-03', mealType: 'Lunch', eatenAt: '2024-01-03T12:00:00+00:00' },
			{ date: '2024-01-03', mealType: 'Dinner', eatenAt: '2024-01-03T19:00:00+00:00' }
		];
		const result = computeMealRegularity(entries);
		expect(result.overallScore).toBeCloseTo(100, 0);
		expect(result.meals.every((m) => m.regularity === 'high')).toBe(true);
	});

	test('random meal times gives lower score', () => {
		const entries = [
			{ date: '2024-01-01', mealType: 'Breakfast', eatenAt: '2024-01-01T06:00:00+00:00' },
			{ date: '2024-01-02', mealType: 'Breakfast', eatenAt: '2024-01-02T10:30:00+00:00' },
			{ date: '2024-01-03', mealType: 'Breakfast', eatenAt: '2024-01-03T07:00:00+00:00' },
			{ date: '2024-01-04', mealType: 'Breakfast', eatenAt: '2024-01-04T11:00:00+00:00' },
			{ date: '2024-01-05', mealType: 'Breakfast', eatenAt: '2024-01-05T08:30:00+00:00' }
		];
		const result = computeMealRegularity(entries);
		expect(result.overallScore).toBeLessThan(100);
	});

	test('takes earliest entry when multiple entries exist for same meal+day', () => {
		const entries = [
			{ date: '2024-01-01', mealType: 'Lunch', eatenAt: '2024-01-01T12:00:00+00:00' },
			{ date: '2024-01-01', mealType: 'Lunch', eatenAt: '2024-01-01T14:00:00+00:00' },
			{ date: '2024-01-02', mealType: 'Lunch', eatenAt: '2024-01-02T12:00:00+00:00' }
		];
		const result = computeMealRegularity(entries);
		const lunch = result.meals.find((m) => m.mealType === 'Lunch');
		expect(lunch?.avgMinute).toBe(12 * 60);
	});

	test('skips entries without eatenAt', () => {
		const entries = [
			{ date: '2024-01-01', mealType: 'Breakfast', eatenAt: null },
			{ date: '2024-01-01', mealType: 'Lunch', eatenAt: '2024-01-01T12:00:00+00:00' }
		];
		const result = computeMealRegularity(entries);
		expect(result.meals).toHaveLength(1);
		expect(result.meals[0].mealType).toBe('Lunch');
	});

	test('returns empty for no timed entries', () => {
		const result = computeMealRegularity([
			{ date: '2024-01-01', mealType: 'Breakfast', eatenAt: null }
		]);
		expect(result.meals).toHaveLength(0);
		expect(result.overallScore).toBe(0);
	});

	test('overallScore is clamped to 0-100', () => {
		const entries = Array.from({ length: 5 }, (_, i) => ({
			date: `2024-01-0${i + 1}`,
			mealType: 'Dinner',
			eatenAt: `2024-01-0${i + 1}T${String(10 + i * 3).padStart(2, '0')}:00:00+00:00`
		}));
		const result = computeMealRegularity(entries);
		expect(result.overallScore).toBeGreaterThanOrEqual(0);
		expect(result.overallScore).toBeLessThanOrEqual(100);
	});
});

File: mobile/shared/src/commonTest/kotlin/com/bissbilanz/analytics/MealRegularityTest.kt
package com.bissbilanz.analytics

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

private fun pad2(n: Int): String = n.toString().padStart(2, '0')

class MealRegularityTest {
    @Test
    fun noEatenAtEntriesReturnsEmpty() {
        val entries =
            listOf(
                RegularityInputEntry(date = "2024-01-01", mealType = "Breakfast", eatenAt = null),
                RegularityInputEntry(date = "2024-01-02", mealType = "Breakfast", eatenAt = null),
            )
        val result = computeMealRegularity(entries)
        assertEquals(emptyList(), result.meals)
        assertEquals(0.0, result.overallScore)
        assertEquals(ConfidenceLevel.INSUFFICIENT, result.confidence)
        assertEquals(0, result.sampleSize)
    }

    @Test
    fun highRegularityWhenStddevBelow30() {
        // Breakfast every day at ~08:00, very consistent (stddev < 30 min)
        val entries =
            (1..14).map { i ->
                val date = "2024-01-${pad2(i)}"
                RegularityInputEntry(date = date, mealType = "Breakfast", eatenAt = "${date}T08:0${i % 10}:00Z")
            }
        val result = computeMealRegularity(entries)
        assertEquals(1, result.meals.size)
        assertEquals("high", result.meals[0].regularity)
        assertTrue(result.meals[0].stddevMinutes < 30)
    }

    @Test
    fun lowRegularityWhenStddevAbove60() {
        // Breakfast at wildly varying times (hour 6 to 13 = 420 to 780 min)
        val times = listOf(6, 13, 7, 12, 6, 13, 7, 12, 6, 13)
        val entries =
            times.mapIndexed { i, hour ->
                val date = "2024-01-${pad2(i + 1)}"
                RegularityInputEntry(date = date, mealType = "Breakfast", eatenAt = "${date}T${pad2(hour)}:00:00Z")
            }
        val result = computeMealRegularity(entries)
        assertEquals(1, result.meals.size)
        assertEquals("low", result.meals[0].regularity)
        assertTrue(result.meals[0].stddevMinutes > 60)
    }

    @Test
    fun overallScoreCalculation() {
        // All meals at exact same time => stddev=0 => score=100
        val entries =
            (1..10).map { i ->
                val date = "2024-01-${pad2(i)}"
                RegularityInputEntry(date = date, mealType = "Lunch", eatenAt = "${date}T12:00:00Z")
            }
        val result = computeMealRegularity(entries)
        assertEquals(100.0, result.overallScore, 0.001)
    }

    @Test
    fun sampleSizeCountsUniqueDates() {
        val entries =
            listOf(
                RegularityInputEntry("2024-01-01", "Breakfast", "2024-01-01T08:00:00Z"),
                RegularityInputEntry("2024-01-01", "Lunch", "2024-01-01T12:00:00Z"),
                RegularityInputEntry("2024-01-02", "Breakfast", "2024-01-02T08:00:00Z"),
            )
        val result = computeMealRegularity(entries)
        assertEquals(2, result.sampleSize)
    }

    @Test
    fun earliestTimePerMealPerDayUsed() {
        // Two breakfast entries on same day — should use earliest (08:00 not 09:00)
        val entries =
            listOf(
                RegularityInputEntry("2024-01-01", "Breakfast", "2024-01-01T09:00:00Z"),
                RegularityInputEntry("2024-01-01", "Breakfast", "2024-01-01T08:00:00Z"),
                RegularityInputEntry("2024-01-02", "Breakfast", "2024-01-02T08:00:00Z"),
            )
        val result = computeMealRegularity(entries)
        assertEquals(1, result.meals.size)
        val meal = result.meals[0]
        // Both days at 08:00 => stddev = 0
        assertEquals(0.0, meal.stddevMinutes, 0.001)
    }
}
